### PR TITLE
Create Holiday Summary report with CSV download support

### PR DESCRIPTION
**Pull request by anarute**
- Add view to display the holiday summary for all active users
- Add service to download the report

The data in this report is currently managed manually with the aggregation of several other reports, so this report is intended to automate this job.

Screen Shot 2022-01-07 at 15 31 32

### CODE REVIEW

**Comment by anarute** [comment]
Depends on #548

File: web/services/downloadHolidaySummary.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+$fp = fopen('php://output', 'wb');
+
+$today = date_create();
+fputcsv($fp, array('Report generated at', $today->format('Y-m-d H:i:s e')), ';');
+fputcsv($fp, array(''), ';');
+
+$weeksLine = array_merge(array('User', 'Hours/day', 'Holidays(days)', 'Holidays(hours)', 'Pending (hours)', 'Planned (hours)', '% planned'), array_keys($usersAndWeeks["weeks"]));
+
+fputcsv($fp, $weeksLine, ';');
```
Comment (jaragunde) [inline web/services/downloadHolidaySummary.php:45]: Just curious, why the semicolon as separator? I think we've been using commas in other places. fputcsv will enclose strings in `""` if they contain the separator character.

File: config/permissions.php [hunk 2]
```diff
@@ -49,6 +49,7 @@
         '/services/syncCalendar.php',
         '/holidayManagement.php',
         '/holidaySummary.php',
+        '/services/downloadHolidaySummary.php',
```
Comment (jaragunde) [inline config/permissions.php:52]: I originally wanted to have services admit a `format` parameter, so we would use `?format=xml` or `?format=csv` to get the same set of data in different formats. We even implemented it in some services. Not a strong opinion, as the codebases is a bit of a mess right now, specially the services layer.

**Review by jaragunde — COMMENTED** [review]
I have some comments. Besides, are you considering to add a way to select the year? That will probably be necessary as we approach the end of 2022.

File: web/services/HolidayService.php [hunk 3]
```diff
@@ -150,6 +150,23 @@ static function formatHours(float $time, float $journey, int $limit): string
...
+    static function getWeeksFromYear($year = NULL): array
+    {
+        $weeks = [];
+        $year = $year ?? date('Y');
+        $first_week = date('W', mktime(0, 0, 0, 1, 1, $year));
+        $last_week = date('W', mktime(0, 0, 0, 12, 31, $year));
+        if ($first_week == "52" || $first_week == "53") {
+            $weeks[($year - 1) . "W" . $first_week] = 0;
+            $first_week = "1";
+        }
```
[the remark this answers is not in the document]
Comment (anarute) [reply web/services/HolidayService.php:167]: Nice catch. Just added a logic for that

**Comment by anarute** [comment]
> I have some comments. Besides, are you considering to add a way to select the year? That will probably be necessary as we approach the end of 2022.

Tbh, I haven't thought about that. I'm hoping that this year we can setup a business inteligence tool (like grafana or metabase) and start doing reports there instead of creating more reports inside PhpReport. I've been experimenting with [Metabase](https://www.metabase.com/) and enjoying it a lot!

File: web/services/downloadHolidaySummary.php (hunk 1 above)
Comment (anarute) [reply web/services/downloadHolidaySummary.php]: No particular reason, probably copied from somewhere else :)

I've just switched to comma.

File: config/permissions.php (hunk 2 above)
Comment (anarute) [reply config/permissions.php:52]: Right, it makes sense to merge them, I just did that.

File: web/services/HolidayService.php [hunk 4]
```diff
@@ -330,4 +330,64 @@ public function syncCalDAVCalendar(array $datesRanges = [])
...
+                'availableDays' => $this::formatHours($summary['availableHours'][$users[$i]->getLogin()], $validJourney, 1),
+                'pendingHours' => round($summary['pendingHours'][$users[$i]->getLogin()], 2),
+                'usedHours' => round($summary['usedHours'][$users[$i]->getLogin()], 2),
+                'percentage' =>  $summary['availableHours'][$users[$i]->getLogin()] ? round(($summary['usedHours'][$users[$i]->getLogin()] / $summary['availableHours'][$users[$i]->getLogin()]) * 100, 2) : 0,
+                'hoursDay' => $validJourney,
+                'holidays' => $leaves,
+            ];
+        }
+        asort($holidays);
+        return [
```
[the remark this answers is not in the document]
Comment (anarute) [reply web/services/HolidayService.php:368]: True, I've moved it to an action: [model/facade/action/GetHolidaySummaryReportAction.php](https://github.com/Igalia/phpreport/pull/550/files#diff-133cddd0577090bfd58ebe45aca8040caffc3e1b1e5168c569cfe44f9c3f25a1)

File: web/services/getHolidaySummary.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+require_once(PHPREPORT_ROOT . "/vendor/autoload.php");
+require_once(PHPREPORT_ROOT . '/util/LoginManager.php');
+
+$csvExport = array_key_exists("format", $_GET) && $_GET["format"] == "csv";
+
+$loginManager = new \LoginManager();
+
+$holidayService = new HolidayService($loginManager);
+
+$usersAndWeeks = $holidayService->retrieveHolidaysSummary();
```
Comment (jaragunde) [inline web/services/getHolidaySummary.php:34]: I passed "2024" as a parameter here, to test the results of a different year, and I'm still getting 2022 results. Can you check why?

**Review by jaragunde — CHANGES_REQUESTED** [review]

File: web/services/getHolidaySummary.php (hunk 5 above)
Comment (anarute) [reply web/services/getHolidaySummary.php]: The year wasn't being passed to the method that generates the weeks per year. Can you try now, please?

I also added support to the service endpoint: `/web/services/getHolidaySummary.php?year=2024`

Comment (jaragunde) [reply web/services/getHolidaySummary.php]: Awesome, thanks!

**Review by jaragunde — APPROVED** [review]
Good job, thank you!

**Pull request merged**